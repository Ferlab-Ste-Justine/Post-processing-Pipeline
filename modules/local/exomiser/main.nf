process EXOMISER {

    label 'process_medium'
 
    input:
    tuple val(meta), path(vcfFile), path(indexFile), path(phenoFile), path(analysisFile)
    path datadir
    val exomiserGenome
    val exomiserDataVersion

    // If specified, the local frequency file path will be inferred from the given path and passed to the exomiser cli.
    // It is expected that the file has a corresponding .tbi index file.
    tuple path(localFrequencyPath), path(localFrequencyIndexPath)

    // If remm/cadd version is specified, remm/cadd reference file(s) path(s) will be inferred from the given filename(s)
    // and passed to the exomiser cli. Each remm/cadd reference file should have a corresponding .tbi index file.
    // Note that, if nextflow adds support for optional paths, one might prefer to pass the full paths explicitly.
    tuple val(remmVersion), val(remmFileName) 
    tuple val(caddVersion), val(caddSnvFileName),val(caddIndelFileName)


    output:
    tuple val(meta), path("results/*vcf.gz")         , optional:true, emit: vcf
    tuple val(meta), path("results/*vcf.gz.tbi")     , optional:true, emit: tbi
    tuple val(meta), path("results/*html")           , optional:true, emit: html
    tuple val(meta), path("results/*json")           , optional:true, emit: json
    tuple val(meta), path("results/*genes.tsv")      , optional:true, emit: genetsv
    tuple val(meta), path("results/*variants.tsv")   , optional:true, emit: variantstsv
    path("versions.yml")            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def applicationPropertiesArgs = task.ext.application_properties_args ?: ''

    def localFrequencyFileArgs = "" 
    if (localFrequencyPath) {
        log.info("Using LOCAL frequency file {}", localFrequencyPath)
        localFrequencyFileArgs = "--exomiser.${exomiserGenome}.local-frequency-path=/`pwd`/${localFrequencyPath}"
    }

    def remmArgs = ""
    if (remmVersion) {
        log.info("Using REMM version {}", remmVersion)
        remmArgs += "--exomiser.remm.version=\"${remmVersion}\""
        remmArgs += " --exomiser.${exomiserGenome}.remm-path=/`pwd`/${datadir}/remm/${remmFileName}"
    }

    def caddArgs = ""
    if (caddVersion) {
        log.info("Using CADD version {}", caddVersion)
        caddArgs += "--cadd.version=\"${caddVersion}\""
        caddArgs += " --exomiser.${exomiserGenome}.cadd-snv-path=/`pwd`/${datadir}/cadd/${caddVersion}/${caddSnvFileName}"
        caddArgs += " --exomiser.${exomiserGenome}.cadd-indel-path=/`pwd`/${datadir}/cadd/${caddVersion}/${caddIndelFileName}"
    }

    def avail_mem = 3072
    if (!task.memory) {
        log.info '[EXOMISER] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
    } else {
        avail_mem = (task.memory.mega*0.8).intValue()
    }

    // Note: specifying the extra options (args) at the beginning because output options are ignored when they are passed at the end.
    """
    #!/bin/bash -eo pipefail

    java -Xmx${avail_mem}M -cp \$( cat /app/jib-classpath-file ) \$( cat /app/jib-main-class-file ) \\
        --vcf ${vcfFile} \\
        --assembly "${exomiserGenome}" \\
        --analysis "${analysisFile}" \\
        --sample ${phenoFile} \\
        --output-format=HTML,JSON,TSV_GENE,TSV_VARIANT,VCF \\
        ${args} \\
        --exomiser.data-directory=/`pwd`/${datadir} \\
        ${localFrequencyFileArgs} \\
        ${remmArgs} \\
        ${caddArgs} \\
        --exomiser.${exomiserGenome}.data-version="${exomiserDataVersion}" \\
        --exomiser.phenotype.data-version="${exomiserDataVersion}" \\
        ${applicationPropertiesArgs}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        "exomiser": "\$(cat /EXOMISER_VERSION.txt)"
    END_VERSIONS
    """

    stub:
    def familyId = meta.familyId
    """
    #!/bin/bash -eo pipefail
    mkdir results
    touch results/${familyId}.exomiser.genes.tsv
    touch results/${familyId}.exomiser.html
    touch results/${familyId}.exomiser.json
    touch results/${familyId}.exomiser.variants.tsv
    touch results/${familyId}.exomiser.vcf.gz
    touch results/${familyId}.exomiser.vcf.gz.tbi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
       "exomiser": "\$(cat /EXOMISER_VERSION.txt)"
    END_VERSIONS
    """
}
