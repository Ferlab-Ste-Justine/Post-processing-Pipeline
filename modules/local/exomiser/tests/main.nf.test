nextflow_process {

    name "Test Process EXOMISER"
    script "modules/local/exomiser/main.nf"
    process "EXOMISER"

    // Using stub mode until we can download a test dataset and write a more elaborated test
    options "-stub"

    tag "local"
    tag "modules"
    tag "exomiser"

    test("Test Process EXOMISER") {
        
        when {
            process {
                """
                input[0] = [ [familyId: "family1"],
                 file("https://github.com/nf-core/test-datasets/raw/modules/data/genomics/homo_sapiens/illumina/gvcf/test.genome.vcf.gz"), 
                 file("assets/exomiser/pheno/family1.yml"),
                 file("assets/exomiser/default_exomiser_WGS_analysis.yml")]
                input[1] = file("data-test/reference/exomiser")
                input[2] = "hg38"
                input[3] = "2402"
                input[4] = "1.7"
                input[5] = "1.3.1"
                """
            }
        }
        then{

            def expected_meta = [familyId: "family1"]
            with(process.out) {
                // vcf channel
                assert vcf.size() == 1
                assert vcf.get(0)[0] == expected_meta
                assert file(vcf.get(0)[1]).name == "family1.splitted-exomiser.vcf.gz"

                // tbi channel
                assert tbi.size() == 1
                assert tbi.get(0)[0] == expected_meta
                assert file(tbi.get(0)[1]).name == "family1.splitted-exomiser.vcf.gz.tbi"

                // html channel
                assert html.size() == 1
                assert html.get(0)[0] == expected_meta
                assert file(html.get(0)[1]).name == "family1.splitted-exomiser.html"

                // json channel
                assert json.size() == 1
                assert json.get(0)[0] == expected_meta
                assert file(json.get(0)[1]).name == "family1.splitted-exomiser.json"

                // genetsv channel
                assert genetsv.size() == 1
                assert genetsv.get(0)[0] == expected_meta
                assert file(genetsv.get(0)[1]).name == "family1.splitted-exomiser.genes.tsv"

                // variantstsv channel
                assert variantstsv.size() == 1
                assert variantstsv.get(0)[0] == expected_meta
                assert file(variantstsv.get(0)[1]).name == "family1.splitted-exomiser.variants.tsv"

                // versions channel
                assert snapshot(versions).match()
            }
        }
    }
}