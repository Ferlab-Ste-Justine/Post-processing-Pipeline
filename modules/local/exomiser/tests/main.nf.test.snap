{
    "Test Process EXOMISER": {
        "content": [
            [
                "versions.yml:md5,30273ee35fb4cdba2d84b2740861902b"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-09-23T15:44:00.728501"
    }
}